File: 23_merge_k_sorted_list/mergeKLists_v2.py
from heapq import heappush, heappop, heapreplace, heapify
class Solution(object):
    def mergeKLists(self, lists):
        """
        :type lists: List[ListNode]
        Heap
        """
        # dummy and node point to the same memory space
        # node is used to operate
        dummy = node = ListNode(0)
        # put all head node of lists into heap
        h = [(n.val, n) for n in lists if n]
        heapify(h)

        while h:
            v, n = h[0]
            # If n has no next, pop it and contunue with next linked list
            if n.next is None:
                heappop(h)
            else:
                heapreplace(h, (n.next.val, n.next))
            node.next = n
            node = node.next

        return dummy.next

File: 23_merge_k_sorted_list/mergeKLists_v2.cpp
#include <queue>
#include <iostream>

using namespace std;

struct ListNode {
    int val;
    ListNode* next;
    ListNode(int x): val(x), next(NULL) {}
};


class Solution {
    public:
        ListNode* mergeKLists(vector<ListNode*>& lists) {
            priority_queue<int> pq;
            ListNode *head, *res = NULL, *dummy = NULL;

            for(int i = 0; i < lists.size(); i++) {
                for(head = lists[i]; head != NULL; head = head->next) {
                    pq.push(head->val);
                }
            }
            while(!pq.empty()) {
                res = new ListNode(pq.top());
                pq.pop();
                res->next = dummy;
                dummy = res;
            }
            return res;
        }
};


int main() {
    int myints[] = {10, 60, 50, 20};

    priority_queue<int, vector<int>, greater<int> > PQ(myints, myints + 4);

    cout << PQ.top() << endl;

    PQ.pop();

    cout << PQ.top() << endl;
    return 0;
}
